common/CommonAPI.cpp

```diff
@@ -0,0 +1,1030 @@
+	for (uint32_t i = 0u; i < imageCount; ++i)
+	{
+		nbl::core::smart_refctd_ptr<nbl::video::IGPUImageView> view[2] = {};
+
+		auto img = swapchain->createImage(i);
+		{
+			nbl::video::IGPUImageView::SCreationParams view_params;
+			view_params.format = img->getCreationParameters().format;
+			view_params.viewType = asset::IImageView<nbl::video::IGPUImage>::ET_2D;
+			view_params.subresourceRange.aspectMask = asset::IImage::EAF_COLOR_BIT;
+			view_params.subresourceRange.baseMipLevel = 0u;
+			view_params.subresourceRange.levelCount = 1u;
+			view_params.subresourceRange.baseArrayLayer = 0u;
+			view_params.subresourceRange.layerCount = 1u;
+			view_params.image = std::move(img);
+
+			view[0] = device->createImageView(std::move(view_params));
+			assert(view[0]);
+		}
+
+		if (useDepth) {
+			nbl::video::IGPUImage::SCreationParams imgParams;
+			imgParams.flags = static_cast<asset::IImage::E_CREATE_FLAGS>(0u);
+			imgParams.type = asset::IImage::ET_2D;
+			imgParams.format = depthFormat;
+			imgParams.extent = { width, height, 1 };
+			imgParams.usage = asset::IImage::E_USAGE_FLAGS::EUF_DEPTH_STENCIL_ATTACHMENT_BIT;
+			imgParams.mipLevels = 1u;
+			imgParams.arrayLayers = 1u;
+			imgParams.samples = asset::IImage::ESCF_1_BIT;
+
+			auto depthImg = device->createImage(std::move(imgParams));
+			auto depthImgMemReqs = depthImg->getMemoryReqs();
+			depthImgMemReqs.memoryTypeBits &= device->getPhysicalDevice()->getDeviceLocalMemoryTypeBits();
+			auto depthImgMem = device->allocate(depthImgMemReqs, depthImg.get());
+
+			nbl::video::IGPUImageView::SCreationParams view_params;
+			view_params.format = depthFormat;
+			view_params.viewType = asset::IImageView<nbl::video::IGPUImage>::ET_2D;
+			view_params.subresourceRange.aspectMask = asset::IImage::EAF_DEPTH_BIT;
+			view_params.subresourceRange.baseMipLevel = 0u;
+			view_params.subresourceRange.levelCount = 1u;
+			view_params.subresourceRange.baseArrayLayer = 0u;
+			view_params.subresourceRange.layerCount = 1u;
+			view_params.image = std::move(depthImg);
+
+			view[1] = device->createImageView(std::move(view_params));
+			assert(view[1]);
+		}
+
+		nbl::video::IGPUFramebuffer::SCreationParams fb_params;
+		fb_params.width = width;
+		fb_params.height = height;
+		fb_params.layers = 1u;
+		fb_params.renderpass = renderpass;
+		fb_params.flags = static_cast<nbl::video::IGPUFramebuffer::E_CREATE_FLAGS>(0);
+		fb_params.attachmentCount = (useDepth) ? 2u : 1u;
+		fb_params.attachments = view;
+
+		fbo->begin()[i] = device->createFramebuffer(std::move(fb_params));
+		assert(fbo->begin()[i]);
+	}
+	return fbo;
+}
+
+bool CommonAPI::createSwapchain(
+	const nbl::core::smart_refctd_ptr<nbl::video::ILogicalDevice>&& device,
+	nbl::video::ISwapchain::SCreationParams& params,
+	uint32_t width, uint32_t height,
+	// nullptr for initial creation, old swapchain for eventual resizes
```

Comment:
this comment makes no sense unless we returned `nbl::core::smart_refctd_ptr<nbl::video::ISwapchain>` instead of `bool`

Reply:
This function takes a reference to the swapchain, places the current value into oldSwapchaib 1and writes the new value into the reference. The first time it's ran it'll be null, then next times it will have the old swapchain

Reply:
Ok i get it now it's a "Reference" to `smart_refctd_ptr" and it is null on the first run.